### PR TITLE
Add mccabe complexity testing

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Add mccabe complexity testing

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
test/

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by sivel** [comment]
When I merged the ignore.txt changes, and re-sorted the file it resulted in a lot of sorting based changes. It looks like only 509 new lines are actually added.

**Comment by mattclay** [comment]
It may be a good idea to first merge a separate PR to sort the ignore file, so lines only need to be added in this PR.

**Comment by mattclay** [comment]
One potential issue with enabling this test is that code which is almost over the limit is likely to either require a refactor (or an added ignore) when minor changes or bug fixes are applied, if those changes push the code over the limit.

**Comment by sivel** [comment]
I don't think there is really anyway around that though. I'm not against adding an ignore in those cases if refactoring is outside the scope. The biggest problem I see is that our ignores are per file, unless it's done in code, rather than `ignore.txt`.

**Comment by mattclay** [comment]
Yeah, there really isn't a way around it. I just wanted to point out one of the downsides to enabling this. I still think it's a good idea to add it.

**Comment by felixfontein** [comment]
@sivel @mattclay how about having a sanity check which enforces that `ignore.txt` is sorted (in a canonical way)? Every time I work on a PR which involves updating `ignore.txt`, and I sort it after adding new entries, there are several changes which are only because of sorting. Also, I suspect that since sorting depends on locale and the sorting program used, different people sorting `ignore.txt` get different orders, which makes the problem worse. So having one canonical and enforced order should be helpful for everyone.

**Review by tremble — APPROVED** [review]
FWIW I like the addition of this test.  It certainly seems to be complaining about some pretty gnarly code in modules/cloud/amazon/.

In my opinion it would be helpful to have a way to manually *lower* the limit during things like initial reviews and reviews of "significant" code rewrites/refactors, this would reduce the risk of a minor change needing significant surgery to avoid an ignore.txt error.  My mind's heading in the direction of "ansible-test sanity --strict ....." which could do things like drop the complexity limit to 15 and testing for unused imports, where gating on them will likely result in more spam in ignore.txt than it's worth, but a human can use that added bit of judgement.

**Comment by sivel** [comment]
> Also, I suspect that since sorting depends on locale and the sorting program used, different people sorting `ignore.txt` get different orders, which makes the problem worse

FWIW, I used python to sort the file, and it seems that python does not use the locale when sorting.  So a helper script, or something exposed through `ansible-test` might be useful.

**Comment by sivel** [comment]
> In my opinion it would be helpful to have a way to manually _lower_ the limit during things like initial reviews

We might be expose an argument of `--max-complexity` when using `sanity` I suppose.

**Comment by mattclay** [comment]
After reviewing some of the files that fail the new test, and thinking more about how this would get used, I'm not sure this is something we should be including in the standard sanity tests.

Up until this point we've done a good job of limiting sanity tests to very low false positives. There is very rarely a reason to add an ignore for a sanity test when adding new code or making changes. For the most part the tests are simply pass/fail and rarely report a fail when they should not.

The mccabe test is different since it provides a score for which there is no correct threshold. This will guarantee the need for ignore entries for valid code that should not, or in some cases, cannot be reasonably refactored to satisfy the test.

This kind of testing seems like it should be purely advisory, and only applied during code review, rather than a hard threshold that must be met or explicitly ignored.

**Comment by sivel** [comment]
> The mccabe test is different since it provides a score for which there is no correct threshold

I suppose this is where it becomes subjective, like many other coding standard tests.  The best you can do is define a sane threshold, just like any other rule used in ensuring a quality of code.

The overall goal with complexity scoring is to add an indicator that the unit of code is too complex to be reliably tested.  Additionally, overly complex code can make it harder to read and reason about.  I know from experience what a high complexity level can do on writing tests, when working with something like `open_url` which weighs in at 31 right now.  I'd argue that they aren't false positives, and are in fact things that can be improved. They should not remain as ignores forever, although I cannot state that someone will ever go back and work on any old ignore.

If we don't have this run and report during CI, no one will ever run it, and we gain nothing.  I'll likely be one the few who ever looks at it, just as I am now.

**Comment by mattclay** [comment]
This may be a good candidate for opt-in sanity test codes as proposed here: https://github.com/ansible/ansible/pull/66920#discussion_r398817968

**Comment by sivel** [comment]
I've updated this PR to make `too-complex` an optional error.  The next commend will be hidden but will include the output, when using `--enable-optional-errors`.

**Comment by sivel** [comment]
```
ERROR: Found 200 pylint issue(s) which need to be resolved:
ERROR: hacking/build_library/build_ansible/command_plugins/plugin_formatter.py:435:0: too-complex: 'process_plugins' is too complex. The McCabe rating is 29
ERROR: lib/ansible/cli/doc.py:144:4: too-complex: 'run' is too complex. The McCabe rating is 30
ERROR: lib/ansible/cli/doc.py:506:4: too-complex: 'add_fields' is too complex. The McCabe rating is 27
ERROR: lib/ansible/cli/doc.py:587:4: too-complex: 'get_man_text' is too complex. The McCabe rating is 23
ERROR: lib/ansible/cli/galaxy.py:499:4: too-complex: '_parse_requirements_file' is too complex. The McCabe rating is 21
ERROR: lib/ansible/cli/playbook.py:68:4: too-complex: 'run' is too complex. The McCabe rating is 24
ERROR: lib/ansible/cli/pull.py:151:4: too-complex: 'run' is too complex. The McCabe rating is 40
ERROR: lib/ansible/cli/scripts/ansible_cli_stub.py:57:0: too-complex: This 'if' is too complex. The McCabe rating is 26
ERROR: lib/ansible/config/manager.py:55:0: too-complex: 'ensure_type' is too complex. The McCabe rating is 28
ERROR: lib/ansible/config/manager.py:425:4: too-complex: 'get_config_value_and_origin' is too complex. The McCabe rating is 27
ERROR: lib/ansible/executor/interpreter_discovery.py:38:0: too-complex: 'discover_interpreter' is too complex. The McCabe rating is 21
ERROR: lib/ansible/executor/module_common.py:696:0: too-complex: 'recursive_finder' is too complex. The McCabe rating is 36
ERROR: lib/ansible/executor/module_common.py:989:0: too-complex: '_find_module_utils' is too complex. The McCabe rating is 31
ERROR: lib/ansible/executor/play_iterator.py:257:4: too-complex: '_get_next_task_from_state' is too complex. The McCabe rating is 36
ERROR: lib/ansible/executor/playbook_executor.py:75:4: too-complex: 'run' is too complex. The McCabe rating is 29
ERROR: lib/ansible/executor/task_executor.py:101:4: too-complex: 'run' is too complex. The McCabe rating is 31
ERROR: lib/ansible/executor/task_executor.py:278:4: too-complex: '_run_loop' is too complex. The McCabe rating is 25
ERROR: lib/ansible/executor/task_executor.py:519:4: too-complex: '_execute' is too complex. The McCabe rating is 60
ERROR: lib/ansible/galaxy/role.py:214:4: too-complex: 'install' is too complex. The McCabe rating is 43
ERROR: lib/ansible/module_utils/basic.py:310:0: too-complex: '_remove_values_conditions' is too complex. The McCabe rating is 21
ERROR: lib/ansible/module_utils/basic.py:1035:4: too-complex: 'set_mode_if_different' is too complex. The McCabe rating is 24
ERROR: lib/ansible/module_utils/basic.py:2194:4: too-complex: 'atomic_move' is too complex. The McCabe rating is 37
ERROR: lib/ansible/module_utils/basic.py:2383:4: too-complex: 'run_command' is too complex. The McCabe rating is 53
ERROR: lib/ansible/module_utils/compat/_selectors2.py:83:0: too-complex: This 'if' is too complex. The McCabe rating is 22
ERROR: lib/ansible/module_utils/compat/_selectors2.py:550:0: too-complex: This 'if' is too complex. The McCabe rating is 23
ERROR: lib/ansible/module_utils/facts/hardware/linux.py:159:4: too-complex: 'get_cpu_facts' is too complex. The McCabe rating is 39
ERROR: lib/ansible/module_utils/facts/hardware/linux.py:642:4: too-complex: 'get_device_facts' is too complex. The McCabe rating is 34
ERROR: lib/ansible/module_utils/facts/network/aix.py:53:4: too-complex: 'get_interfaces_info' is too complex. The McCabe rating is 27
ERROR: lib/ansible/module_utils/facts/network/linux.py:99:4: too-complex: 'get_interfaces_info' is too complex. The McCabe rating is 44
ERROR: lib/ansible/module_utils/facts/system/service_mgr.py:65:4: too-complex: 'collect' is too complex. The McCabe rating is 21
ERROR: lib/ansible/module_utils/facts/virtual/linux.py:36:4: too-complex: 'get_virtual_facts' is too complex. The McCabe rating is 59
ERROR: lib/ansible/module_utils/facts/virtual/sunos.py:33:4: too-complex: 'get_virtual_facts' is too complex. The McCabe rating is 28
ERROR: lib/ansible/module_utils/six/__init__.py:752:0: too-complex: This 'if' is too complex. The McCabe rating is 23
ERROR: lib/ansible/module_utils/urls.py:178:0: too-complex: This 'if' is too complex. The McCabe rating is 40
ERROR: lib/ansible/module_utils/urls.py:414:0: too-complex: This 'if' is too complex. The McCabe rating is 23
ERROR: lib/ansible/module_utils/urls.py:807:4: too-complex: 'get_ca_certs' is too complex. The McCabe rating is 23
ERROR: lib/ansible/module_utils/urls.py:1096:4: too-complex: 'open' is too complex. The McCabe rating is 31
ERROR: lib/ansible/modules/apt.py:438:0: too-complex: 'package_status' is too complex. The McCabe rating is 24
ERROR: lib/ansible/modules/apt.py:1014:0: too-complex: 'main' is too complex. The McCabe rating is 41
ERROR: lib/ansible/modules/blockinfile.py:193:0: too-complex: 'main' is too complex. The McCabe rating is 34
ERROR: lib/ansible/modules/copy.py:332:0: too-complex: 'chown_recursive' is too complex. The McCabe rating is 29
ERROR: lib/ansible/modules/copy.py:492:0: too-complex: 'main' is too complex. The McCabe rating is 72
ERROR: lib/ansible/modules/cron.py:535:0: too-complex: 'main' is too complex. The McCabe rating is 43
ERROR: lib/ansible/modules/dnf.py:879:4: too-complex: 'ensure' is too complex. The McCabe rating is 95
ERROR: lib/ansible/modules/file.py:662:0: too-complex: 'ensure_symlink' is too complex. The McCabe rating is 22
ERROR: lib/ansible/modules/file.py:765:0: too-complex: 'ensure_hardlink' is too complex. The McCabe rating is 28
ERROR: lib/ansible/modules/find.py:355:0: too-complex: 'main' is too complex. The McCabe rating is 25
ERROR: lib/ansible/modules/get_url.py:422:0: too-complex: 'main' is too complex. The McCabe rating is 43
ERROR: lib/ansible/modules/git.py:1043:0: too-complex: 'main' is too complex. The McCabe rating is 40
ERROR: lib/ansible/modules/lineinfile.py:253:0: too-complex: 'present' is too complex. The McCabe rating is 48
ERROR: lib/ansible/modules/pip.py:584:0: too-complex: 'main' is too complex. The McCabe rating is 39
ERROR: lib/ansible/modules/service.py:435:4: too-complex: 'get_service_tools' is too complex. The McCabe rating is 27
ERROR: lib/ansible/modules/service.py:669:4: too-complex: 'service_enable' is too complex. The McCabe rating is 56
ERROR: lib/ansible/modules/service.py:1163:4: too-complex: 'service_enable' is too complex. The McCabe rating is 23
ERROR: lib/ansible/modules/service_facts.py:95:4: too-complex: 'gather_services' is too complex. The McCabe rating is 23
ERROR: lib/ansible/modules/stat.py:432:0: too-complex: 'main' is too complex. The McCabe rating is 24
ERROR: lib/ansible/modules/systemd.py:321:0: too-complex: 'main' is too complex. The McCabe rating is 45
ERROR: lib/ansible/modules/sysvinit.py:117:0: too-complex: 'main' is too complex. The McCabe rating is 51
ERROR: lib/ansible/modules/unarchive.py:293:4: too-complex: 'is_unarchived' is too complex. The McCabe rating is 72
ERROR: lib/ansible/modules/uri.py:586:0: too-complex: 'main' is too complex. The McCabe rating is 38
ERROR: lib/ansible/modules/user.py:578:4: too-complex: 'create_user_useradd' is too complex. The McCabe rating is 28
ERROR: lib/ansible/modules/user.py:713:4: too-complex: 'modify_user_usermod' is too complex. The McCabe rating is 36
ERROR: lib/ansible/modules/user.py:989:4: too-complex: 'ssh_key_gen' is too complex. The McCabe rating is 21
ERROR: lib/ansible/modules/user.py:1260:4: too-complex: 'modify_user' is too complex. The McCabe rating is 35
ERROR: lib/ansible/modules/user.py:1488:4: too-complex: 'modify_user' is too complex. The McCabe rating is 26
ERROR: lib/ansible/modules/user.py:1661:4: too-complex: 'modify_user' is too complex. The McCabe rating is 24
ERROR: lib/ansible/modules/user.py:1796:4: too-complex: 'create_user' is too complex. The McCabe rating is 30
ERROR: lib/ansible/modules/user.py:1897:4: too-complex: 'modify_user_usermod' is too complex. The McCabe rating is 31
ERROR: lib/ansible/modules/user.py:2649:4: too-complex: 'modify_user' is too complex. The McCabe rating is 21
ERROR: lib/ansible/modules/user.py:2861:0: too-complex: 'main' is too complex. The McCabe rating is 29
ERROR: lib/ansible/modules/wait_for.py:456:0: too-complex: 'main' is too complex. The McCabe rating is 58
ERROR: lib/ansible/modules/yum.py:946:4: too-complex: 'install' is too complex. The McCabe rating is 33
ERROR: lib/ansible/modules/yum.py:1241:4: too-complex: 'latest' is too complex. The McCabe rating is 39
ERROR: lib/ansible/modules/yum.py:1447:4: too-complex: 'ensure' is too complex. The McCabe rating is 31
ERROR: lib/ansible/modules/yum.py:1570:4: too-complex: 'run' is too complex. The McCabe rating is 21
ERROR: lib/ansible/playbook/base.py:338:4: too-complex: 'get_validated_value' is too complex. The McCabe rating is 25
ERROR: lib/ansible/playbook/block.py:299:4: too-complex: '_get_parent_attribute' is too complex. The McCabe rating is 23
ERROR: lib/ansible/playbook/conditional.py:102:4: too-complex: '_check_conditional' is too complex. The McCabe rating is 27
ERROR: lib/ansible/playbook/helpers.py:84:0: too-complex: 'load_list_of_tasks' is too complex. The McCabe rating is 49
ERROR: lib/ansible/playbook/included_file.py:61:4: too-complex: 'process_include_results' is too complex. The McCabe rating is 40
ERROR: lib/ansible/playbook/play_context.py:201:4: too-complex: 'set_task_and_variable_override' is too complex. The McCabe rating is 35
ERROR: lib/ansible/playbook/task.py:168:4: too-complex: 'preprocess_data' is too complex. The McCabe rating is 21
ERROR: lib/ansible/plugins/action/__init__.py:785:4: too-complex: '_execute_module' is too complex. The McCabe rating is 38
ERROR: lib/ansible/plugins/action/copy.py:226:4: too-complex: '_copy_file' is too complex. The McCabe rating is 23
ERROR: lib/ansible/plugins/action/copy.py:401:4: too-complex: 'run' is too complex. The McCabe rating is 33
ERROR: lib/ansible/plugins/action/fetch.py:37:4: too-complex: 'run' is too complex. The McCabe rating is 37
ERROR: lib/ansible/plugins/action/pause.py:76:4: too-complex: 'run' is too complex. The McCabe rating is 40
ERROR: lib/ansible/plugins/action/template.py:28:4: too-complex: 'run' is too complex. The McCabe rating is 30
ERROR: lib/ansible/plugins/callback/__init__.py:167:4: too-complex: '_get_diff' is too complex. The McCabe rating is 22
ERROR: lib/ansible/plugins/connection/ssh.py:561:4: too-complex: '_build_command' is too complex. The McCabe rating is 21
ERROR: lib/ansible/plugins/connection/ssh.py:783:4: too-complex: '_bare_run' is too complex. The McCabe rating is 48
ERROR: lib/ansible/plugins/inventory/script.py:91:4: too-complex: 'parse' is too complex. The McCabe rating is 21
ERROR: lib/ansible/plugins/loader.py:612:4: too-complex: '_find_plugin_legacy' is too complex. The McCabe rating is 21
ERROR: lib/ansible/plugins/lookup/subelements.py:97:4: too-complex: 'run' is too complex. The McCabe rating is 21
ERROR: lib/ansible/plugins/strategy/__init__.py:439:4: too-complex: '_process_pending_results' is too complex. The McCabe rating is 75
ERROR: lib/ansible/plugins/strategy/__init__.py:1109:4: too-complex: '_execute_meta' is too complex. The McCabe rating is 28
ERROR: lib/ansible/plugins/strategy/free.py:71:4: too-complex: 'run' is too complex. The McCabe rating is 36
ERROR: lib/ansible/plugins/strategy/linear.py:198:4: too-complex: 'run' is too complex. The McCabe rating is 44
ERROR: lib/ansible/template/__init__.py:395:4: too-complex: '__getitem__' is too complex. The McCabe rating is 23
ERROR: lib/ansible/template/__init__.py:689:4: too-complex: 'template' is too complex. The McCabe rating is 25
ERROR: lib/ansible/template/__init__.py:952:4: too-complex: 'do_template' is too complex. The McCabe rating is 23
ERROR: lib/ansible/template/safe_eval.py:31:0: too-complex: 'safe_eval' is too complex. The McCabe rating is 21
ERROR: lib/ansible/utils/plugin_docs.py:43:0: too-complex: '_process_versions_and_dates' is too complex. The McCabe rating is 24
ERROR: lib/ansible/utils/plugin_docs.py:106:0: too-complex: 'add_fragments' is too complex. The McCabe rating is 21
ERROR: lib/ansible/vars/manager.py:146:4: too-complex: 'get_vars' is too complex. The McCabe rating is 64
ERROR: test/integration/targets/incidental_script_inventory_vmware_inventory/vmware_inventory.py:429:4: too-complex: 'instances_to_inventory' is too complex. The McCabe rating is 29
ERROR: test/integration/targets/incidental_script_inventory_vmware_inventory/vmware_inventory.py:565:4: too-complex: 'facts_from_proplist' is too complex. The McCabe rating is 22
ERROR: test/integration/targets/incidental_script_inventory_vmware_inventory/vmware_inventory.py:688:4: too-complex: '_process_object_types' is too complex. The McCabe rating is 28
ERROR: test/integration/targets/var_precedence/ansible-var-precedence-check.py:197:4: too-complex: 'build' is too complex. The McCabe rating is 34
ERROR: test/lib/ansible_test/_data/sanity/code-smell/shebang.py:11:0: too-complex: 'main' is too complex. The McCabe rating is 22
ERROR: test/lib/ansible_test/_data/sanity/import/importer.py:7:0: too-complex: 'main' is too complex. The McCabe rating is 77
ERROR: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:908:4: too-complex: '_validate_docs' is too complex. The McCabe rating is 35
ERROR: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:1263:4: too-complex: '_validate_required_if' is too complex. The McCabe rating is 26
ERROR: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:1410:4: too-complex: '_validate_argument_spec' is too complex. The McCabe rating is 128
ERROR: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:2052:4: too-complex: 'validate' is too complex. The McCabe rating is 23
ERROR: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:2212:0: too-complex: 'run' is too complex. The McCabe rating is 21
ERROR: test/lib/ansible_test/_internal/classification.py:58:0: too-complex: 'categorize_changes' is too complex. The McCabe rating is 22
ERROR: test/lib/ansible_test/_internal/classification.py:372:4: too-complex: '_classify_common' is too complex. The McCabe rating is 49
ERROR: test/lib/ansible_test/_internal/classification.py:654:4: too-complex: '_classify_ansible' is too complex. The McCabe rating is 48
ERROR: test/lib/ansible_test/_internal/core_ci.py:60:4: too-complex: '__init__' is too complex. The McCabe rating is 22
ERROR: test/lib/ansible_test/_internal/delegation.py:223:0: too-complex: 'delegate_docker' is too complex. The McCabe rating is 25
ERROR: test/lib/ansible_test/_internal/delegation.py:397:0: too-complex: 'delegate_remote' is too complex. The McCabe rating is 21
ERROR: test/lib/ansible_test/_internal/executor.py:645:0: too-complex: 'command_windows_integration' is too complex. The McCabe rating is 25
ERROR: test/lib/ansible_test/_internal/executor.py:980:0: too-complex: 'command_integration_filtered' is too complex. The McCabe rating is 34
ERROR: test/lib/ansible_test/_internal/import_analysis.py:29:0: too-complex: 'get_python_module_utils_imports' is too complex. The McCabe rating is 26
ERROR: test/lib/ansible_test/_internal/sanity/__init__.py:79:0: too-complex: 'command_sanity' is too complex. The McCabe rating is 31
ERROR: test/lib/ansible_test/_internal/sanity/__init__.py:248:4: too-complex: '__init__' is too complex. The McCabe rating is 32
ERROR: test/lib/ansible_test/_internal/target.py:353:0: too-complex: 'analyze_integration_target_dependencies' is too complex. The McCabe rating is 31
ERROR: test/lib/ansible_test/_internal/units/__init__.py:54:0: too-complex: 'command_units' is too complex. The McCabe rating is 22
ERROR: test/support/integration/plugins/inventory/foreman.py:211:4: too-complex: '_populate' is too complex. The McCabe rating is 25
ERROR: test/support/integration/plugins/lookup/rabbitmq.py:119:4: too-complex: 'run' is too complex. The McCabe rating is 21
ERROR: test/support/integration/plugins/module_utils/compat/ipaddress.py:1807:4: too-complex: '_ip_int_from_string' is too complex. The McCabe rating is 23
ERROR: test/support/integration/plugins/module_utils/docker/common.py:760:0: too-complex: 'compare_generic' is too complex. The McCabe rating is 26
ERROR: test/support/integration/plugins/module_utils/ec2.py:239:0: too-complex: 'get_aws_connection_info' is too complex. The McCabe rating is 31
ERROR: test/support/integration/plugins/module_utils/k8s/raw.py:236:4: too-complex: 'perform_action' is too complex. The McCabe rating is 48
ERROR: test/support/integration/plugins/module_utils/net_tools/nios/api.py:242:4: too-complex: 'run' is too complex. The McCabe rating is 44
ERROR: test/support/integration/plugins/module_utils/net_tools/nios/api.py:485:4: too-complex: 'get_object_ref' is too complex. The McCabe rating is 27
ERROR: test/support/integration/plugins/module_utils/vmware.py:281:0: too-complex: 'gather_vm_facts' is too complex. The McCabe rating is 30
ERROR: test/support/integration/plugins/modules/aws_s3.py:664:0: too-complex: 'main' is too complex. The McCabe rating is 54
ERROR: test/support/integration/plugins/modules/azure_rm_functionapp.py:234:4: too-complex: 'exec_module' is too complex. The McCabe rating is 25
ERROR: test/support/integration/plugins/modules/azure_rm_mariadbdatabase.py:158:4: too-complex: 'exec_module' is too complex. The McCabe rating is 21
ERROR: test/support/integration/plugins/modules/azure_rm_mariadbserver.py:220:4: too-complex: 'exec_module' is too complex. The McCabe rating is 28
ERROR: test/support/integration/plugins/modules/azure_rm_resource.py:315:4: too-complex: 'exec_module' is too complex. The McCabe rating is 25
ERROR: test/support/integration/plugins/modules/azure_rm_resource_info.py:336:4: too-complex: 'exec_module' is too complex. The McCabe rating is 23
ERROR: test/support/integration/plugins/modules/azure_rm_storageaccount.py:504:4: too-complex: 'update_account' is too complex. The McCabe rating is 28
ERROR: test/support/integration/plugins/modules/azure_rm_webapp.py:550:4: too-complex: 'exec_module' is too complex. The McCabe rating is 60
ERROR: test/support/integration/plugins/modules/azure_rm_webappslot.py:458:4: too-complex: 'exec_module' is too complex. The McCabe rating is 54
ERROR: test/support/integration/plugins/modules/cloudformation.py:644:0: too-complex: 'main' is too complex. The McCabe rating is 36
ERROR: test/support/integration/plugins/modules/ec2.py:991:0: too-complex: 'create_instances' is too complex. The McCabe rating is 68
ERROR: test/support/integration/plugins/modules/ec2.py:1402:0: too-complex: 'startstop_instances' is too complex. The McCabe rating is 25
ERROR: test/support/integration/plugins/modules/ec2_eni.py:372:0: too-complex: 'modify_eni' is too complex. The McCabe rating is 21
ERROR: test/support/integration/plugins/modules/ec2_group.py:592:0: too-complex: 'get_target_from_rule' is too complex. The McCabe rating is 24
ERROR: test/support/integration/plugins/modules/ec2_group.py:999:0: too-complex: 'get_diff_final_resource' is too complex. The McCabe rating is 23
ERROR: test/support/integration/plugins/modules/ec2_group.py:1137:0: too-complex: 'main' is too complex. The McCabe rating is 33
ERROR: test/support/integration/plugins/modules/ec2_instance.py:1180:0: too-complex: 'build_top_level_options' is too complex. The McCabe rating is 23
ERROR: test/support/integration/plugins/modules/ec2_instance.py:1673:0: too-complex: 'main' is too complex. The McCabe rating is 24
ERROR: test/support/integration/plugins/modules/ec2_vpc_net.py:366:0: too-complex: 'main' is too complex. The McCabe rating is 40
ERROR: test/support/integration/plugins/modules/hcloud_server.py:363:4: too-complex: '_update_server' is too complex. The McCabe rating is 22
ERROR: test/support/integration/plugins/modules/lvg.py:143:0: too-complex: 'main' is too complex. The McCabe rating is 31
ERROR: test/support/integration/plugins/modules/mongodb_user.py:355:0: too-complex: 'main' is too complex. The McCabe rating is 28
ERROR: test/support/integration/plugins/modules/mysql_db.py:304:0: too-complex: 'db_dump' is too complex. The McCabe rating is 24
ERROR: test/support/integration/plugins/modules/mysql_db.py:459:0: too-complex: 'main' is too complex. The McCabe rating is 37
ERROR: test/support/integration/plugins/modules/mysql_user.py:367:0: too-complex: 'user_mod' is too complex. The McCabe rating is 34
ERROR: test/support/integration/plugins/modules/mysql_user.py:688:0: too-complex: 'main' is too complex. The McCabe rating is 27
ERROR: test/support/integration/plugins/modules/openssl_certificate.py:1811:4: too-complex: 'assertonly' is too complex. The McCabe rating is 36
ERROR: test/support/integration/plugins/modules/openssl_certificate.py:2536:0: too-complex: 'main' is too complex. The McCabe rating is 39
ERROR: test/support/integration/plugins/modules/openssl_csr.py:712:4: too-complex: '_check_csr' is too complex. The McCabe rating is 24
ERROR: test/support/integration/plugins/modules/openssl_csr.py:805:4: too-complex: '_generate_csr' is too complex. The McCabe rating is 23
ERROR: test/support/integration/plugins/modules/openssl_csr.py:898:4: too-complex: '_check_csr' is too complex. The McCabe rating is 42
ERROR: test/support/integration/plugins/modules/postgresql_db.py:514:0: too-complex: 'main' is too complex. The McCabe rating is 31
ERROR: test/support/integration/plugins/modules/postgresql_privs.py:658:4: too-complex: 'manipulate_privs' is too complex. The McCabe rating is 30
ERROR: test/support/integration/plugins/modules/postgresql_privs.py:921:0: too-complex: 'main' is too complex. The McCabe rating is 36
ERROR: test/support/integration/plugins/modules/postgresql_query.py:241:0: too-complex: 'main' is too complex. The McCabe rating is 26
ERROR: test/support/integration/plugins/modules/postgresql_user.py:368:0: too-complex: 'user_alter' is too complex. The McCabe rating is 37
ERROR: test/support/integration/plugins/modules/postgresql_user.py:806:0: too-complex: 'main' is too complex. The McCabe rating is 26
ERROR: test/support/integration/plugins/modules/s3_bucket.py:183:0: too-complex: 'create_or_update_bucket' is too complex. The McCabe rating is 63
ERROR: test/support/integration/plugins/modules/synchronize.py:380:0: too-complex: 'main' is too complex. The McCabe rating is 56
ERROR: test/support/integration/plugins/modules/ufw.py:324:0: too-complex: 'main' is too complex. The McCabe rating is 54
ERROR: test/support/integration/plugins/modules/vmware_guest.py:1048:4: too-complex: 'configure_cpu_and_memory' is too complex. The McCabe rating is 37
ERROR: test/support/integration/plugins/modules/vmware_guest.py:1300:4: too-complex: 'configure_hardware_params' is too complex. The McCabe rating is 30
ERROR: test/support/integration/plugins/modules/vmware_guest.py:1503:4: too-complex: 'configure_network' is too complex. The McCabe rating is 25
ERROR: test/support/integration/plugins/modules/vmware_guest.py:1778:4: too-complex: 'customize_vm' is too complex. The McCabe rating is 28
ERROR: test/support/integration/plugins/modules/vmware_guest.py:2159:4: too-complex: 'select_datastore' is too complex. The McCabe rating is 22
ERROR: test/support/integration/plugins/modules/vmware_guest.py:2333:4: too-complex: 'deploy_vm' is too complex. The McCabe rating is 48
ERROR: test/support/integration/plugins/modules/vmware_guest.py:2586:4: too-complex: 'reconfigure_vm' is too complex. The McCabe rating is 27
ERROR: test/support/integration/plugins/modules/vmware_host_hyperthreading.py:122:4: too-complex: 'ensure' is too complex. The McCabe rating is 24
ERROR: test/support/integration/plugins/modules/xml.py:810:0: too-complex: 'main' is too complex. The McCabe rating is 23
ERROR: test/support/integration/plugins/modules/zabbix_host.py:449:4: too-complex: 'update_host' is too complex. The McCabe rating is 24
ERROR: test/support/integration/plugins/modules/zabbix_host.py:600:4: too-complex: 'check_all_properties' is too complex. The McCabe rating is 43
ERROR: test/support/integration/plugins/modules/zabbix_host.py:778:0: too-complex: 'main' is too complex. The McCabe rating is 59
ERROR: test/support/network-integration/collections/ansible_collections/ansible/netcommon/plugins/action/net_put.py:36:4: too-complex: 'run' is too complex. The McCabe rating is 23
ERROR: test/support/network-integration/collections/ansible_collections/ansible/netcommon/plugins/filter/ipaddr.py:484:0: too-complex: 'ipaddr' is too complex. The McCabe rating is 39
ERROR: test/support/network-integration/collections/ansible_collections/ansible/netcommon/plugins/filter/ipaddr.py:789:0: too-complex: 'ipsubnet' is too complex. The McCabe rating is 26
ERROR: test/support/network-integration/collections/ansible_collections/ansible/netcommon/plugins/filter/network.py:83:0: too-complex: 'parse_cli' is too complex. The McCabe rating is 36
ERROR: test/support/network-integration/collections/ansible_collections/ansible/netcommon/plugins/module_utils/compat/ipaddress.py:1889:4: too-complex: '_ip_int_from_string' is too complex. The McCabe rating is 23
ERROR: test/support/network-integration/collections/ansible_collections/ansible/netcommon/plugins/modules/cli_config.py:241:0: too-complex: 'run' is too complex. The McCabe rating is 27
ERROR: test/support/network-integration/collections/ansible_collections/cisco/ios/plugins/modules/ios_config.py:404:0: too-complex: 'main' is too complex. The McCabe rating is 27
ERROR: test/support/windows-integration/plugins/action/win_copy.py:332:4: too-complex: 'run' is too complex. The McCabe rating is 25
ERROR: test/units/module_utils/basic/test_deprecate_warn.py:27:4: ansible-deprecated-date: Expired date ('ansible.builtin:2020-03-10') found in call to Display.deprecated or AnsibleModule.deprecate
ERROR: test/units/module_utils/facts/system/distribution/test_distribution_version.py:28:0: too-complex: 'test_distribution_version' is too complex. The McCabe rating is 25
```

File: test/lib/ansible_test/_data/requirements/sanity.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 cryptography
 jinja2
+mccabe
```
Comment (mattclay) [inline test/lib/ansible_test/_data/requirements/sanity.txt]: The sanity test files are now split up by sanity test, so this belongs here now:

https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_data/requirements/sanity.pylint.txt

Comment (sivel) [reply test/lib/ansible_test/_data/requirements/sanity.txt]: done.

**Review by mattclay — APPROVED** [review]

**Review by mattclay — CHANGES_REQUESTED** [review]
Needs a changelog fragment. I'd also freeze the plugin version in the constraints file.

Other than that, it looks ready to merge.

**Comment by mattclay** [comment]
CI failure are unrelated to changes.

**Review by mattclay — APPROVED** [review]

**Pull request merged**